### PR TITLE
Added an interface to the data reader to return a Conduit node

### PR DESCRIPTION
**Pull request by bvanessen**
directly from the data reader to the data coordinator using a new
fetch function interface.  Added a data_packer class that uses the
infrastructure from the HDF5 data reader to extract data fields from
Conduit nodes and pack them into the I/O buffers.  Updated the HDF5
data reader to use the new Conduit fetch function interface.  Added a
new field in the data reader class to indicate that concrete data
readers support the new interface.

### CODE REVIEW

**Comment by bvanessen** [comment]
Passes CI https://lc.llnl.gov/gitlab/vanessen/lbann/-/pipelines/157119

File: src/data_coordinator/unit_test/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+set_full_path(THIS_DIR_MPI_CATCH2_TEST_FILES
+  data_coordinator_HDF5_hrrl_public_api.cpp
+  )
+
+set(LBANN_SEQ_CATCH2_TEST_FILES
+  "${LBANN_SEQ_CATCH2_TEST_FILES}"
+  "${THIS_DIR_SEQ_CATCH2_TEST_FILES}" PARENT_SCOPE)
+set(LBANN_MPI_CATCH2_TEST_FILES
+  "${LBANN_MPI_CATCH2_TEST_FILES}"
+  "${THIS_DIR_MPI_CATCH2_TEST_FILES}" PARENT_SCOPE)
```
Comment (benson31) [inline src/data_coordinator/unit_test/CMakeLists.txt:13]:
```suggestion
set_full_path(THIS_DIR_MPI_CATCH2_TEST_FILES
  data_coordinator_HDF5_hrrl_public_api.cpp
  )

set(LBANN_MPI_CATCH2_TEST_FILES
  "${LBANN_MPI_CATCH2_TEST_FILES}"
  "${THIS_DIR_MPI_CATCH2_TEST_FILES}" PARENT_SCOPE)
```
Also add the license.

File: include/lbann/data_coordinator/data_packer.hpp [hunk 2]
```diff
@@ -0,0 +1,53 @@
+////////////////////////////////////////////////////////////////////////////////
+// Copyright (c) 2014-2021, Lawrence Livermore National Security, LLC.
```
Comment (benson31) [inline include/lbann/data_coordinator/data_packer.hpp:2]:
```suggestion
// Copyright (c) 2014-2022, Lawrence Livermore National Security, LLC.
```

File: src/data_coordinator/data_packer.cpp [hunk 3]
```diff
@@ -0,0 +1,198 @@
+////////////////////////////////////////////////////////////////////////////////
+// Copyright (c) 2014-2021, Lawrence Livermore National Security, LLC.
```
Comment (benson31) [inline src/data_coordinator/data_packer.cpp:2]:
```suggestion
// Copyright (c) 2014-2022, Lawrence Livermore National Security, LLC.
```

File: src/data_coordinator/unit_test/data_coordinator_HDF5_hrrl_public_api.cpp [hunk 4]
```diff
@@ -0,0 +1,281 @@
+////////////////////////////////////////////////////////////////////////////////
+// Copyright (c) 2014-2021, Lawrence Livermore National Security, LLC.
```
Comment (benson31) [inline src/data_coordinator/unit_test/data_coordinator_HDF5_hrrl_public_api.cpp:2]:
```suggestion
// Copyright (c) 2014-2022, Lawrence Livermore National Security, LLC.
```

File: src/proto/data_coordinator.proto [hunk 5]
```diff
@@ -0,0 +1,46 @@
+////////////////////////////////////////////////////////////////////////////////
+// Copyright (c) 2014-2019, Lawrence Livermore National Security, LLC.
```
Comment (benson31) [inline src/proto/data_coordinator.proto:2]:
```suggestion
// Copyright (c) 2014-2022, Lawrence Livermore National Security, LLC.
```

**Review by benson31 — COMMENTED** [review]
Review from 50000 ft. I'm still reading it more closely for C++ content.

**Review by benson31 — APPROVED** [review]
I'm good with this now

**Pull request merged**